File: sdk-dotnet/src/Flux/Classifier.cs
// Clasificación de errores del handler.
// Contrato normativo: specification/04-errors.md §2
//
// Este fichero es el punto donde el protocolo se encuentra con la realidad operativa del
// ecosistema. Todo lo demás en el SDK es mecánica; esto es política — y por eso la
// política es un parámetro, no una constante.

using System.Net.Http;
using System.Net.Sockets;

namespace Flux;

/// <summary>
/// Qué hacer con un error que no encaja en ninguna regla conocida — 04-errors.md §2.1.
/// </summary>
/// <remarks>
/// Es un tipo propio y no un <see cref="ErrorClass"/> porque "retryable acotado" NO es una
/// clase del protocolo: es la clase RETRYABLE con un presupuesto distinto. Meterlo en
/// <see cref="ErrorClass"/> contaminaría el valor que acaba escrito en <c>dlqreason</c>.
/// <para>
/// <see cref="RetryableBounded"/> vale 0 para que <c>default(ClassifierOptions)</c>
/// signifique el default de la spec.
/// </para>
/// </remarks>
public enum UnknownPolicy
{
    /// <summary>
    /// El default de la spec: RETRYABLE con presupuesto reducido
    /// (<see cref="ClassifierOptions.UnknownRetryBudget"/>, 2 entregas) en vez de los 6
    /// completos. Un transitorio se recupera en el segundo intento; un sistemático llega a
    /// la DLQ en ~30 s sin atascar la cola.
    /// </summary>
    RetryableBounded = 0,

    /// <summary>
    /// A la DLQ sin gastar reintentos. Falla rápido, pero un hipo de red manda a la DLQ un
    /// evento perfectamente válido y alguien lo reproduce a mano cada mañana.
    /// </summary>
    Permanent = 1,

    /// <summary>
    /// Backoff completo, 51 minutos. Elígelo solo si vuestras dependencias internas tienen
    /// hipos frecuentes y podéis asumir que un modo de fallo nuevo atasque la cola y se
    /// amplifique con cada mensaje que falle igual.
    /// </summary>
    Retryable = 2,
}

/// <summary>La política de clasificación de errores del consumidor.</summary>
public sealed record ClassifierOptions
{
    /// <summary>
    /// Qué hacer con un error que no encaja en ninguna regla conocida. Ver
    /// <see cref="UnknownPolicy"/>.
    /// </summary>
    public UnknownPolicy UnknownErrorPolicy { get; init; } = UnknownPolicy.RetryableBounded;

    /// <summary>
    /// Entregas máximas de un error desconocido cuando la política es
    /// <see cref="UnknownPolicy.RetryableBounded"/>. Incluye la primera entrega, así que
    /// 2 = un reintento. Cero significa el default de la spec
    /// (<see cref="Classifier.DefaultUnknownRetryBudget"/>).
    /// </summary>
    /// <remarks>
    /// NO se traduce a <c>max_deliver</c> del consumidor: eso es por consumidor, no por
    /// mensaje, y recortaría también los reintentos de los RETRYABLE reconocidos. Viaja en
    /// <see cref="Classification.MaxAttempts"/> y lo aplica el runtime a ese error concreto
    /// — 04-errors.md §2.1.
    /// </remarks>
    public int UnknownRetryBudget { get; init; }

    /// <summary>
    /// Un timeout, ¿es "el mundo va lento" o "esta operación no cabe en la ventana"?
    /// </summary>
    /// <remarks>
    /// El default es <see cref="ErrorClass.Retryable"/>: un timeout suele indicar saturación
    /// transitoria. Si vuestros timeouts son casi siempre consultas que nunca van a
    /// terminar, <see cref="ErrorClass.Permanent"/> evita reintentar lo imposible.
    /// </remarks>
    public ErrorClass TimeoutPolicy { get; init; } = ErrorClass.Retryable;

    /// <summary>
    /// Reglas propias, evaluadas antes que todo lo demás salvo los errores que ya declaran
    /// su clase. Devolver <see langword="null"/> cede a la siguiente regla.
    /// </summary>
    public IReadOnlyList<Func<Exception, Classification?>> Rules { get; init; } =
        Array.Empty<Func<Exception, Classification?>>();
}

/// <summary>
/// Traduce un error cualquiera a una de las tres clases del protocolo.
/// </summary>
/// <remarks>
/// El runtime del consumidor usa el resultado así:
/// <list type="bullet">
/// <item><description><see cref="ErrorClass.Retryable"/> → <c>nak(RetryAfter ?? backoff canónico)</c></description></item>
/// <item><description><see cref="ErrorClass.Permanent"/> → <c>term()</c> + publicar en <c>dlq.&lt;subject&gt;</c></description></item>
/// <item><description><see cref="ErrorClass.Poison"/> → <c>term()</c> + publicar en <c>dlq.&lt;subject&gt;</c> + alerta inmediata</description></item>
/// </list>
/// El orden de evaluación es deliberado: lo más específico primero y el default al final.
/// Esa última línea es la decisión de política de verdad.
/// </remarks>
public sealed class Classifier
{
    /// <summary>
    /// Entregas que gasta un error desconocido bajo la política acotada. Incluye la primera
    /// entrega, así que 2 = un reintento.
    /// </summary>
    public const int DefaultUnknownRetryBudget = 2;

    /// <summary>
    /// Status HTTP que merecen reintento — 04-errors.md §1.1.
    /// </summary>
    /// <remarks>
    /// Nótese qué NO está aquí: 400, 403, 404 y 422 son PERMANENT. Reintentarlos es gastar
    /// 51 minutos para obtener exactamente la misma respuesta.
    /// </remarks>
    public static readonly IReadOnlySet<int> RetryableHttpStatus = new HashSet<int> { 429, 502, 503, 504 };

    /// <summary>
    /// Errores de socket inequívocamente transitorios, con su nombre POSIX.
    /// </summary>
    /// <remarks>
    /// La clasificación se hace por <see cref="SocketError"/>, que es el mecanismo
    /// idiomático de .NET, y NO por coincidencia de subcadenas sobre el mensaje del error
    /// — que es justo lo que invita a hacer la lista de códigos de <c>protocol.json</c> y
    /// lo que 04-errors.md §1.1 prohíbe explícitamente.
    /// <para>
    /// Esto además inmuniza al SDK contra el bug real que produjo el port literal de la
    /// lista de Node: en Windows los códigos de libuv llevan prefijo <c>WSA</c>
    /// (<c>WSAECONNRESET</c>), así que el mismo corte de red se clasificaba PERMANENT en
    /// Windows y RETRYABLE en Linux. <see cref="SocketError"/> ya normaliza esa diferencia
    /// en el BCL: el enum es el mismo en las dos plataformas.
    /// </para>
    /// <para>
    /// El nombre POSIX se emite como <see cref="Classification.Code"/> —y no
    /// <c>"ConnectionReset"</c>— para que las métricas de un consumidor .NET se puedan
    /// agregar con las de uno de Node o Go: es el mismo hecho operativo y merece la misma
    /// etiqueta.
    /// </para>
    /// <para>
    /// Nótese qué NO está aquí: <see cref="SocketError.HostNotFound"/>,
    /// <see cref="SocketError.NoData"/> y <see cref="SocketError.NoRecovery"/> son
    /// respuestas DEFINITIVAS del resolutor ("no existe"), no temporales
    /// (<see cref="SocketError.TryAgain"/>, el <c>EAI_AGAIN</c> de la spec). Reintentar un
    /// nombre que no existe son 51 minutos para llegar al mismo sitio.
    /// </para>
    /// </remarks>
    public static readonly IReadOnlyDictionary<SocketError, string> TransientSocketErrors =
        new Dictionary<SocketError, string>
        {
            [SocketError.ConnectionReset] = "ECONNRESET",
            [SocketError.ConnectionRefused] = "ECONNREFUSED",
            [SocketError.ConnectionAborted] = "ECONNABORTED",
            [SocketError.TimedOut] = "ETIMEDOUT",
            [SocketError.HostUnreachable] = "EHOSTUNREACH",
            [SocketError.NetworkUnreachable] = "ENETUNREACH",
            [SocketError.NetworkDown] = "ENETDOWN",
            [SocketError.NetworkReset] = "ENETRESET",
            [SocketError.HostDown] = "EHOSTDOWN",
            [SocketError.TryAgain] = "EAI_AGAIN",
        };

    /// <summary>Profundidad máxima al recorrer la cadena de errores. Corta ciclos.</summary>
    private const int MaxChainDepth = 32;

    private readonly ErrorClass _unknownClass;
    private readonly int _unknownBudget;
    private readonly ErrorClass _timeoutClass;
    private readonly Func<Exception, Classification?>[] _rules;

    /// <summary>Construye un clasificador con la política indicada.</summary>
    /// <param name="options">
    /// Política. <see langword="null"/> significa los defaults de la spec: desconocido →
    /// RETRYABLE acotado a 2 entregas, timeout → RETRYABLE.
    /// </param>
    public Classifier(ClassifierOptions? options = null)
    {
        var opts = options ?? new ClassifierOptions();

        _unknownClass = opts.UnknownErrorPolicy == UnknownPolicy.Permanent
            ? ErrorClass.Permanent
            : ErrorClass.Retryable;

        // Solo la política acotada impone un tope propio; las otras dos dejan mandar al
        // max_deliver del consumidor.
        _unknownBudget = opts.UnknownErrorPolicy == UnknownPolicy.RetryableBounded
            ? (opts.UnknownRetryBudget > 0 ? opts.UnknownRetryBudget : DefaultUnknownRetryBudget)
            : 0;

        _timeoutClass = opts.TimeoutPolicy == ErrorClass.Permanent
            ? ErrorClass.Permanent
            : ErrorClass.Retryable;

        // Copia: la política no cambia bajo los pies si el llamante muta su lista.
        _rules = opts.Rules.ToArray();
    }

    /// <summary>Clasificador con los defaults de la spec.</summary>
    public static Classifier Default { get; } = new();

    /// <summary>
    /// Entregas que el runtime concede a ESTE error concreto.
    /// </summary>
    /// <remarks>
    /// <c>budget = min(consumerMaxDeliver, classification.MaxAttempts &gt; 0 ? MaxAttempts :
    /// consumerMaxDeliver)</c> — 04-errors.md §2.1.
    /// <para>
    /// Es lo que permite que un error DESCONOCIDO se agote en 2 entregas (~30 s) mientras un
    /// RETRYABLE RECONOCIDO (ECONNRESET, HTTP 503) conserva sus 6 entregas y sus 51 minutos:
    /// el tope viaja en la clasificación del error, no en el <c>max_deliver</c> del
    /// consumidor, que es por consumidor y recortaría a los dos por igual.
    /// </para>
    /// <para>
    /// Vive aquí y no dentro del bucle de consumo para que la regla más importante del
    /// runtime se pueda probar sin levantar un broker.
    /// </para>
    /// </remarks>
    /// <param name="consumerMaxDeliver">El <c>max_deliver</c> del consumidor.</param>
    /// <param name="classification">La clasificación del error que acaba de fallar.</param>
    public static int EffectiveBudget(int consumerMaxDeliver, Classification classification)
    {
        ArgumentNullException.ThrowIfNull(classification);
        return classification.MaxAttempts > 0
            ? Math.Min(consumerMaxDeliver, classification.MaxAttempts)
            : consumerMaxDeliver;
    }

    /// <summary>Clasifica un error del handler.</summary>
    public Classification Classify(Exception? error)
    {
        if (error is null)
        {
            // No debería ocurrir —el runtime solo clasifica errores— pero devolver una
            // clase falsa sería peor que devolver algo inerte.
            return new Classification(ErrorClass.Permanent, "NIL_ERROR");
        }

        // 1. Un error tipado de flux siempre gana: la aplicación sabe más que el SDK.
        //    Se busca en toda la cadena, así que un RetryableException envuelto por una
        //    capa intermedia sigue clasificándose bien — cosa que el `instanceof` de Node,
        //    que solo mira el error de arriba, no consigue.
        var classified = FindInChain<ClassifiedException>(error);
        if (classified is not null)
        {
            var result = new Classification(classified.FluxClass, classified.FluxCode);
            return classified.FluxClass == ErrorClass.Retryable
                ? result with { RetryAfter = FindInChain<IRetryAfterError>(error)?.RetryAfter }
                : result;
        }

        // 2. Reglas de la aplicación.
        foreach (var rule in _rules)
        {
            var r = rule(error);
            if (r is not null)
            {
                return r;
            }
        }

        // 3. Status HTTP: la señal más fiable que da una dependencia.
        //    Primero el contrato explícito del SDK, y si no, el del BCL: desde .NET 5
        //    HttpRequestException lleva el StatusCode, así que la mayoría de aplicaciones
        //    no tienen que implementar nada.
        var status = FindInChain<IHttpStatusError>(error)?.HttpStatus
                     ?? (int?)FindInChain<HttpRequestException>(error)?.StatusCode;
        if (status is not null)
        {
            var code = "HTTP_" + status.Value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            if (!RetryableHttpStatus.Contains(status.Value))
            {
                return new Classification(ErrorClass.Permanent, code);
            }

            return new Classification(ErrorClass.Retryable, code)
            {
                RetryAfter = FindInChain<IRetryAfterError>(error)?.RetryAfter,
            };
        }

        // 4. Errores de red: transitorios por semántica, no por lista de códigos.
        var socket = FindInChain<SocketException>(error);
        if (socket is not null && TransientSocketErrors.TryGetValue(socket.SocketErrorCode, out var posix))
        {
            return new Classification(ErrorClass.Retryable, posix);
        }

        // 5. Fallo de transporte HTTP sin respuesta: la petición nunca llegó a obtener un
        //    status (DNS, conexión, TLS). Es "el mundo ahora mismo", no el evento.
        if (FindInChain<HttpRequestException>(error) is not null)
        {
            return new Classification(ErrorClass.Retryable, "HTTP_TRANSPORT");
        }

        // 6. Timeouts — política configurable.
        //
        //    ⚠️ Fricción de .NET: OperationCanceledException NO distingue "se agotó el
        //    plazo" de "el llamante canceló". HttpClient lanza TaskCanceledException en
        //    ambos casos. Se tratan todos como timeout porque en un handler de flux la
        //    cancelación viene del apagado del proceso, y entonces no confirmar el mensaje
        //    y dejar que se reentregue es exactamente lo correcto.
        if (FindInChain<TimeoutException>(error) is not null ||
            FindInChain<OperationCanceledException>(error) is not null)
        {
            return new Classification(_timeoutClass, "TIMEOUT");
        }

        // 7. Lo desconocido. Aquí se decide el comportamiento del ecosistema ante lo que
        //    nadie previó. El default acotado da al transitorio una segunda oportunidad sin
        //    regalarle 51 minutos de cola al sistemático — y el tope viaja en la
        //    clasificación, no en max_deliver, para no recortar los reintentos de los
        //    RETRYABLE reconocidos — 04-errors.md §2.1.
        return new Classification(_unknownClass, "UNKNOWN") { MaxAttempts = _unknownBudget };
    }

    /// <summary>
    /// Busca el primer error de tipo <typeparamref name="T"/> en la cadena.
    /// </summary>
    /// <remarks>
    /// Equivale al <c>errors.As</c> de Go y atraviesa tanto
    /// <see cref="Exception.InnerException"/> como las ramas de
    /// <see cref="AggregateException"/>, que es como llegan los fallos de un
    /// <c>Task.WhenAll</c> dentro de un handler.
    /// </remarks>
    private static T? FindInChain<T>(Exception? error, int depth = 0)
        where T : class
    {
        if (error is null || depth > MaxChainDepth)
        {
            return null;
        }

        if (error is T match)
        {
            return match;
        }

        if (error is AggregateException aggregate)
        {
            foreach (var inner in aggregate.InnerExceptions)
            {
                var found = FindInChain<T>(inner, depth + 1);
                if (found is not null)
                {
                    return found;
                }
            }

            return null;
        }

        return FindInChain<T>(error.InnerException, depth + 1);
    }
}

File: sdk-java/src/main/java/com/flux/Classifier.java
/*
 * Clasificacion de errores del handler.
 * Contrato normativo: specification/04-errors.md §2
 *
 * Este fichero es el punto donde el protocolo se encuentra con la realidad operativa del
 * ecosistema. Todo lo demas en el SDK es mecanica; esto es politica — y por eso la
 * politica es un parametro, no una constante.
 */
package com.flux;

import com.flux.FluxErrors.Classification;
import java.io.InterruptedIOException;
import java.net.ConnectException;
import java.net.NoRouteToHostException;
import java.net.PortUnreachableException;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.net.http.HttpTimeoutException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.TimeoutException;
import java.util.function.Function;

/**
 * Traduce un error cualquiera a una de las tres clases del protocolo.
 *
 * <p>El runtime del consumidor usa el resultado asi:
 *
 * <pre>
 *   RETRYABLE  → msg.nakWithDelay(retryAfter ?: backoff canonico)
 *   PERMANENT  → msg.term() + publicar en dlq.&lt;subject&gt; con dlqattempts = intento
 *   POISON     → msg.term() + publicar en dlq.&lt;subject&gt; + alerta inmediata
 * </pre>
 */
public final class Classifier {

    // ─── Politicas ───────────────────────────────────────────────────────────

    /**
     * Que hacer con un error que no encaja en ninguna regla conocida — 04-errors.md §2.1.
     */
    public enum UnknownErrorPolicy {
        /**
         * A la DLQ sin gastar reintentos. Falla rapido, pero un hipo de red manda a la DLQ
         * un evento perfectamente valido y alguien lo reproduce a mano cada manana.
         */
        PERMANENT,

        /**
         * Backoff completo: 51 minutos. Solo si vuestras dependencias tienen hipos muy
         * frecuentes y podeis asumir que un modo de fallo nuevo atasque la cola — y se
         * amplifique con cada mensaje que falle igual.
         */
        RETRYABLE,

        /**
         * DEFAULT. Reintenta con presupuesto reducido ({@link #DEFAULT_UNKNOWN_RETRY_BUDGET}
         * entregas) en vez de las 6 completas.
         *
         * <p>Domina a las otras dos: un transitorio desconocido se recupera en el segundo
         * intento y un sistematico desconocido llega a la DLQ en ~30 s sin atascar la cola.
         * Cuesta 30 segundos de latencia sobre los errores genuinamente permanentes y a
         * cambio elimina los dos modos de fallo. No es un punto medio, es estrictamente
         * mejor — 04-errors.md §2.1.
         */
        RETRYABLE_BOUNDED
    }

    /** Un timeout, ¿es "el mundo va lento" o "esta operacion no cabe en la ventana"? */
    public enum TimeoutPolicy {
        /** Evita reintentar lo imposible, si vuestros timeouts son consultas que nunca terminan. */
        PERMANENT,
        /** DEFAULT: un timeout suele indicar saturacion transitoria. */
        RETRYABLE
    }

    /**
     * Entregas maximas para un error desconocido bajo {@link UnknownErrorPolicy#RETRYABLE_BOUNDED}.
     * Incluye la primera entrega, asi que 2 = un reintento — protocol.json
     * {@code errors.unknownRetryBudget}.
     */
    public static final int DEFAULT_UNKNOWN_RETRY_BUDGET = 2;

    /**
     * Status HTTP que merecen reintento — 04-errors.md §1.1.
     *
     * <p>Notese que NO esta aqui: 400, 403, 404 y 422 son PERMANENT. Reintentarlos es
     * gastar 51 minutos para obtener exactamente la misma respuesta.
     */
    public static final Set<Integer> RETRYABLE_HTTP_STATUS = Set.of(429, 502, 503, 504);

    // ─── Contratos que la aplicacion puede implementar ───────────────────────

    /**
     * Error que sabe decir con que status HTTP fallo una dependencia. El status es la
     * senal mas fiable que da una dependencia sobre si merece la pena reintentar.
     *
     * <p>⚠️ Divergencia con Node, y es de fondo: alli el clasificador hurga en
     * {@code err.status}, {@code err.statusCode} y {@code err.response.status} porque en
     * JavaScript cualquier objeto puede tener cualquier propiedad. En Java eso exigiria
     * reflexion sobre nombres de campo, que es adivinar; el contrato se hace explicito.
     * Implementa esta interfaz —o usa {@link HttpException}— y el clasificador la
     * encuentra recorriendo la cadena de causas.
     */
    public interface HttpStatusAware {
        int httpStatus();
    }

    /** Error que sabe cuanto pide esperar la dependencia (cabecera {@code Retry-After}). */
    public interface RetryAfterAware {
        Duration retryAfter();
    }

    /**
     * Implementacion de conveniencia de {@link HttpStatusAware}, para no obligar a cada
     * aplicacion a escribir la suya.
     *
     * <pre>{@code
     * if (response.statusCode() >= 400) {
     *     throw new Classifier.HttpException(response.statusCode(), "POST /v1/charges", retryAfter);
     * }
     * }</pre>
     */
    public static class HttpException extends RuntimeException implements HttpStatusAware, RetryAfterAware {
        private static final long serialVersionUID = 1L;

        private final int status;
        private final transient Duration retryAfter;

        public HttpException(int status, String message) {
            this(status, message, null, null);
        }

        public HttpException(int status, String message, Duration retryAfter) {
            this(status, message, retryAfter, null);
        }

        public HttpException(int status, String message, Duration retryAfter, Throwable cause) {
            super(message == null ? "HTTP " + status : "HTTP " + status + ": " + message, cause);
            this.status = status;
            this.retryAfter = retryAfter;
        }

        @Override
        public int httpStatus() {
            return status;
        }

        /** {@code null} si la dependencia no anuncio {@code Retry-After}. */
        @Override
        public Duration retryAfter() {
            return retryAfter;
        }
    }

    /**
     * Regla de clasificacion de la aplicacion. Devuelve {@link Optional#empty()} para
     * ceder al resto de la cadena. Se evaluan antes que todo lo demas salvo los errores
     * que ya declaran su clase.
     */
    @FunctionalInterface
    public interface Rule extends Function<Throwable, Optional<Classification>> {
    }

    // ─── Errores transitorios, por SEMANTICA ─────────────────────────────────

    /**
     * Tipos de excepcion que significan "la red no esta disponible o se interrumpio", y
     * "la resolucion de nombres fallo de forma temporal" — 04-errors.md §1.1.
     *
     * <p>⚠️ La lista de {@code protocol.json} ({@code ECONNRESET}, {@code EAI_AGAIN}…) son
     * codigos de libuv y NO son normativos: la clasificacion se define por semantica. En
     * Java el mecanismo idiomatico es el TIPO de la excepcion, que el propio JDK ya elige
     * segun el errno. Nunca se hace matching de substrings sobre el mensaje: el mensaje de
     * {@code SocketException} lo escribe el sistema operativo y difiere entre Windows y
     * Linux, que es justo el bug que la spec documenta (un port literal de la lista de
     * Node clasificaba PERMANENT en Windows y RETRYABLE en Linux el mismo corte de red).
     *
     * <p>El orden importa: {@link ConnectException}, {@link NoRouteToHostException} y
     * {@link PortUnreachableException} extienden {@link SocketException}, asi que la
     * superclase va la ultima para que el {@code code} de la metrica sea el tipo mas
     * especifico.
     *
     * <p>Cobertura por categoria de la spec:
     * <ul>
     *   <li>Conexion rechazada / dependencia arrancando → {@link ConnectException}</li>
     *   <li>Ruta inalcanzable → {@link NoRouteToHostException},
     *       {@link PortUnreachableException}</li>
     *   <li>Conexion reseteada, tuberia rota, red caida → {@link SocketException}</li>
     *   <li>Resolucion de nombres → {@link UnknownHostException}</li>
     * </ul>
     */
    private static final List<Class<? extends Throwable>> TRANSIENT_NETWORK = List.of(
            ConnectException.class,
            NoRouteToHostException.class,
            PortUnreachableException.class,
            UnknownHostException.class,
            SocketException.class);

    /**
     * Tipos que significan "se agoto el plazo".
     *
     * <p>{@link SocketTimeoutException} extiende {@link InterruptedIOException} y NO
     * {@link SocketException}, asi que no lo captura la lista de red: la separacion entre
     * "la red fallo" y "la operacion no cabio en la ventana" —que la spec trata con
     * politicas distintas— la da el propio arbol de tipos del JDK.
     */
    private static final List<Class<? extends Throwable>> TIMEOUTS = List.of(
            SocketTimeoutException.class,
            HttpTimeoutException.class,
            TimeoutException.class,
            InterruptedIOException.class);

    // ─── Opciones ────────────────────────────────────────────────────────────

    /**
     * Politica de clasificacion del consumidor.
     *
     * <p>Builder mutable por la misma razon que {@link Envelope.BuildEventInput}: todos
     * los campos son opcionales y tienen un default de la spec.
     */
    public static final class ClassifierOptions {
        private UnknownErrorPolicy unknownErrorPolicy = UnknownErrorPolicy.RETRYABLE_BOUNDED;
        private int unknownRetryBudget = DEFAULT_UNKNOWN_RETRY_BUDGET;
        private TimeoutPolicy timeoutPolicy = TimeoutPolicy.RETRYABLE;
        private final List<Rule> rules = new ArrayList<>();

        public ClassifierOptions unknownErrorPolicy(UnknownErrorPolicy policy) {
            this.unknownErrorPolicy = policy;
            return this;
        }

        /** Entregas maximas de un error desconocido bajo RETRYABLE_BOUNDED. Minimo 1. */
        public ClassifierOptions unknownRetryBudget(int budget) {
            if (budget < 1) {
                throw new IllegalArgumentException("el presupuesto minimo es 1 entrega (la inicial)");
            }
            this.unknownRetryBudget = budget;
            return this;
        }

        public ClassifierOptions timeoutPolicy(TimeoutPolicy policy) {
            this.timeoutPolicy = policy;
            return this;
        }

        /** Regla propia, evaluada antes que todo lo demas salvo los errores tipados. */
        public ClassifierOptions addRule(Rule rule) {
            this.rules.add(rule);
            return this;
        }
    }

    // ─── Clasificador ────────────────────────────────────────────────────────

    private final UnknownErrorPolicy unknownErrorPolicy;
    private final int unknownRetryBudget;
    private final TimeoutPolicy timeoutPolicy;
    private final List<Rule> rules;

    /** Clasificador con los defaults de la spec. */
    public Classifier() {
        this(new ClassifierOptions());
    }

    public Classifier(ClassifierOptions options) {
        ClassifierOptions o = options != null ? options : new ClassifierOptions();
        this.unknownErrorPolicy = o.unknownErrorPolicy;
        this.unknownRetryBudget = o.unknownRetryBudget;
        this.timeoutPolicy = o.timeoutPolicy;
        // Copia defensiva: la politica no debe poder cambiar bajo los pies del consumidor
        // una vez creado. Es lo mismo que hace el clasificador de Go con su slice de reglas.
        this.rules = List.copyOf(o.rules);
    }

    /** Clasificador con los defaults de la spec: desconocido → RETRYABLE acotado (2). */
    public static final Classifier DEFAULT = new Classifier();

    /**
     * Clasifica un error.
     *
     * <p>El orden de evaluacion es deliberado: lo mas especifico primero y el default al
     * final. Esa ultima linea es la decision de politica de verdad.
     */
    public Classification classify(Throwable error) {
        if (error == null) {
            // No deberia ocurrir —el runtime solo clasifica errores— pero devolver una
            // clase falsa seria peor que devolver algo inerte.
            return new Classification(ErrorClass.PERMANENT, "NIL_ERROR");
        }

        // 1. Un error tipado de flux siempre gana: la aplicacion sabe mas que el SDK.
        Optional<FluxErrors.FluxException> typed = FluxErrors.asClassified(error);
        if (typed.isPresent()) {
            FluxErrors.FluxException fe = typed.get();
            Duration retryAfter = fe instanceof FluxErrors.RetryableException re
                    ? re.retryAfter().orElse(null)
                    : null;
            return new Classification(fe.errorClass(), fe.code(), retryAfter, null);
        }

        // 2. Reglas de la aplicacion.
        for (Rule rule : rules) {
            Optional<Classification> result = rule.apply(error);
            if (result.isPresent()) {
                return result.get();
            }
        }

        // 3. Status HTTP: la senal mas fiable que da una dependencia.
        Optional<HttpStatusAware> http = FluxErrors.findCause(error, HttpStatusAware.class);
        if (http.isPresent()) {
            int status = http.get().httpStatus();
            boolean retryable = RETRYABLE_HTTP_STATUS.contains(status);
            if (!retryable) {
                return new Classification(ErrorClass.PERMANENT, "HTTP_" + status);
            }
            Duration retryAfter = FluxErrors.findCause(error, RetryAfterAware.class)
                    .map(RetryAfterAware::retryAfter)
                    .orElse(null);
            return new Classification(ErrorClass.RETRYABLE, "HTTP_" + status, retryAfter, null);
        }

        // 4. Errores de sistema: red y DNS son transitorios por definicion.
        //
        // UnknownHostException entra aqui, y con una imprecision que conviene conocer: el
        // JDK usa el mismo tipo para "el nombre no existe" (NXDOMAIN, permanente) y para
        // "el resolutor no pudo contestar ahora" (SERVFAIL, transitorio). La spec pide
        // reintentar solo el segundo. Java no expone la diferencia —no hay equivalente al
        // *net.DNSError.IsTemporary de Go—, asi que se reintenta en ambos casos: el coste
        // de reintentar un host inexistente esta acotado por max_deliver, y el de NO
        // reintentar un SERVFAIL es tirar un evento bueno a la DLQ. Ver README §"Fricciones".
        for (Class<? extends Throwable> type : TRANSIENT_NETWORK) {
            Optional<? extends Throwable> found = FluxErrors.findCause(error, type);
            if (found.isPresent()) {
                return new Classification(ErrorClass.RETRYABLE, found.get().getClass().getSimpleName());
            }
        }

        // 5. Timeouts — politica configurable.
        for (Class<? extends Throwable> type : TIMEOUTS) {
            if (FluxErrors.findCause(error, type).isPresent()) {
                ErrorClass timeoutClass = timeoutPolicy == TimeoutPolicy.PERMANENT
                        ? ErrorClass.PERMANENT
                        : ErrorClass.RETRYABLE;
                return new Classification(timeoutClass, "TIMEOUT");
            }
        }

        // 6. Lo desconocido. Aqui se decide el comportamiento del ecosistema ante lo que
        //    nadie previo. El default acotado da al transitorio una segunda oportunidad sin
        //    regalarle 51 minutos de cola al sistematico — 04-errors.md §2.1.
        return switch (unknownErrorPolicy) {
            case PERMANENT -> new Classification(ErrorClass.PERMANENT, "UNKNOWN");
            case RETRYABLE -> new Classification(ErrorClass.RETRYABLE, "UNKNOWN");
            case RETRYABLE_BOUNDED -> Classification.retryableBounded("UNKNOWN", unknownRetryBudget);
        };
    }
}
